### PR TITLE
feat: add loading indicator when changing password and saving profile

### LINKED ISSUE #67: Add loading indicators
## Description
As a user I would like to see the status of various tasks, like changing password, saving profile,etc
I would like to see a loading indicator when the process is running, instead of just a snack bar once the process has finished.
The following are the tasks where i would like to see it:
1. Change password
2. Saving profile.
3. (will add more incase I find any, feel free to add other such processes in the comments below)

### Update [Required]
- [ ] Add loading indicators

## Definition of Done
- [ ] All of the required items are completed.
- [x] Approval by 1 mentor.

### LINKED ISSUE #94: Add Loading indicator when a task is created and marked as complete
**Is your feature request related to a problem? Please describe.**
Yes.  

1. When a user creates a task, after clicking create button, there is a delay of 1-2 seconds for a response from the app which is not user-freindly. 
2. Similar situation when a user marks a task as completed it stands still for 2sec.

**Describe the solution you'd like**
When a user creates a task, it has to show a loading indicator.
Similarly when a user checks the task item as completed it has to show a loading indicator.

**Additional context**
Current version of Task created: 
Record-2020-06-11-17-49-47-92746
Current version of Task marked as complete:
Record-2020-06-11-17-51-03-92746

### PR DESCRIPTION
**Pull request by techno-disaster**
### Description
Add loading indicator on screen to help user understand what's happening.
If a user has a slow internet connection, the changing password request can take long and without a loading indicator he/she might think the apps stuck. Loading indicators will tell that the process is currently going on  
Fixes #67
Fixes #94

### Flutter Channel:
- [x] I have used the Flutter Beta channel on my local machine 

### Type of Change:
**Delete irrelevant options.**

- User Interface

**Code/Quality Assurance Only**
- New feature (non-breaking change which adds functionality pre-approved by mentors)

### How Has This Been Tested?

Physical device. Check gif below
ezgif com-video-to-gif

### Checklist:
**Delete irrelevant options.**

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials
- [x] I have commented my code or provided relevant documentation, particularly in hard-to-understand areas

**Code/Quality Assurance Only**
- [x] My changes generate no new warnings

### CODE REVIEW

**Comment by techno-disaster** [comment]
@anitab-org/qa-team can someone test this? 
Steps : 
Change password, a loading indicator like in the gif should load up, until the snackbar comes up
Change something in Profile and save profile, a loading indicator should show up until the snacbar comes

File: lib/screens/home/pages/profile/profile_page.dart [hunk 1]
```diff
@@ -220,4 +222,12 @@ class _ProfilePageState extends State<ProfilePage> {
       ),
     );
   }
+
+  Future<void> showloader(BuildContext context) {
```
Comment (bartekpacia) [inline lib/screens/home/pages/profile/profile_page.dart]: Please change the name to _showProgressIndicator()

File: lib/screens/settings/settings_screen.dart [hunk 2]
```diff
@@ -74,16 +76,15 @@ class SettingsScreen extends StatelessWidget {
             ),
           ],
         );
-      }
+      },
     );
   }
 
-  Future<void> _showChangePasswordDialog(BuildContext context) async {
+  Future<void> _showChangePasswordDialog(BuildContext theContext) async {
```
Comment (bartekpacia) [inline lib/screens/settings/settings_screen.dart]: If we want to name `BuildContext`s like this, let's use `topContext` instead of `theContext`.

**Review by bartekpacia — CHANGES_REQUESTED** [review]

**Comment by bartekpacia** [comment]
Great job overall👍

File: lib/screens/settings/settings_screen.dart (hunk 2 above)
Comment (techno-disaster) [reply lib/screens/settings/settings_screen.dart]: sure. will change asap

**Review by bartekpacia — DISMISSED** [review]
Looks good.
EDIT: Looks I had a second reviewing screen opened in the second browser tab🤦‍♀️

**Comment by techno-disaster** [comment]
> Looks good.

Still have to make some changes.

**Comment by techno-disaster** [comment]
#### Found one more page which can use loader and cleaned he code a bit

ezgif com-video-to-gif(1)

**Comment by techno-disaster** [comment]
I had to remove ```final``` from the bloc because I wanted to set the message to null, after we show the snackbar. without this the loader runs fine for the first time, but if the user hits send request again, then as state.message is not null anymore it wont enter the ifcase, and no snackbar will be shown + continuous loading.

**Review by bartekpacia — DISMISSED** [review]
LGTM

File: lib/screens/settings/settings_screen.dart (hunk 2 above)
Comment (techno-disaster) [reply lib/screens/settings/settings_screen.dart]: Done

File: lib/screens/home/pages/profile/profile_page.dart (hunk 1 above)
Comment (techno-disaster) [reply lib/screens/home/pages/profile/profile_page.dart]: Done

**Comment by techno-disaster** [comment]
@isabelcosta can you review this too? 👀 Thanks

**Review by robotjellyzone — DISMISSED** [review]
Looks good to me @Techno-Disaster ! just want to know why did you downgraded some of the versions in pubspec file ?

**Comment by techno-disaster** [comment]
> Looks good to me @Techno-Disaster ! just want to know why did you downgraded some of the versions in pubspec file ?

The pubspec.lock is an auto generated file. We store all our dependencies in pubspec.yaml file.

**Comment by techno-disaster** [comment]
Working on #94  (It needs loading indicators as well). Lets review it after that's pushed too

**Comment by techno-disaster** [comment]
Added indicators to tasks page.
ezgif com-video-to-gif(1)

**Comment by techno-disaster** [comment]
@anitab-org/qa-team  can someone give this a quick test? 
Thank You

File: lib/screens/send_request/bloc/send_request_state.dart [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import 'package:equatable/equatable.dart';
 
 abstract class SendRequestState extends Equatable {
-  final String message;
+  String message;
```
Comment (Gaurav192) [inline lib/screens/send_request/bloc/send_request_state.dart]: Removing final is not a proper solution.
I would suggest to use a resetEvent which yields initalState once message has been listened to.

Comment (techno-disaster) [reply lib/screens/send_request/bloc/send_request_state.dart]: @Gaurav192 thanks for the suggestion, updated the pr can you give it a quick look? Thanks

File: lib/screens/send_request/bloc/send_request_bloc.dart [hunk 4]
```diff
@@ -28,5 +28,13 @@ class SendRequestBloc extends Bloc<SendRequestEvent, SendRequestState> {
         yield InitialSendRequestState(message: failure.message);
       }
     }
+    if (event is ResetSnackbarMessage) {
+      try {
```
Comment (Gaurav192) [inline lib/screens/send_request/bloc/send_request_bloc.dart]: try catch is not required here.
yield initialState;  is enough

Comment (techno-disaster) [reply lib/screens/send_request/bloc/send_request_bloc.dart]: @Gaurav192 done :)

File: lib/screens/home/pages/relation/relation_page.dart [hunk 5]
```diff
@@ -163,8 +165,8 @@ class _RelationPageState extends State<RelationPage> {
                     GestureDetector(
                       onTap: () {
                         if (!task.isDone) {
-                          context.toast("hey");
```
Comment (HaripriyaB) [inline lib/screens/home/pages/relation/relation_page.dart:166]: @Techno-Disaster I think taking off this line was supposed to be for another issue which we discussed in last meeting right?

Comment (techno-disaster) [reply lib/screens/home/pages/relation/relation_page.dart:166]: @HaripriyaB yes, but IMHO it's just 1 small line and it doesn't really make any sense to have a whole new Issue for it.

Comment (HaripriyaB) [reply lib/screens/home/pages/relation/relation_page.dart:166]: yes that makes sense...

**Review by HaripriyaB — DISMISSED** [review]

**Review by bartekpacia — DISMISSED** [review]
nice!

**Comment by robotjellyzone** [comment]
> @anitab-org/qa-team can someone give this a quick test?
> Thank You

I have tested this PR @Techno-Disaster as per your request . The needed feature is working as can be seen below :-
testing

**Comment by robotjellyzone** [comment]
But  I don't know why relation page isn't working as it can be seen from the following gif that all other pages are working but i am unable to see the relation page as its causing the app to crash ... with the following message on logcat  -->

 `INFO: 2020-06-13 16:38:29.914830: Transition { currentState: RelationPageLoading, event: RelationPageShowed, nextState: RelationPageFailure }`

relation-page-not-working

**Comment by robotjellyzone** [comment]
May be the changes done in the relation page needs to be reviewed again 🤔 @Techno-Disaster ?

**Review by robotjellyzone — COMMENTED** [review]
May be the changes done in the relation page needs to be reviewed again 🤔 @Techno-Disaster ?

**Comment by techno-disaster** [comment]
@robotjellyzone thanks for reporting this, this happened when the user was not in a relationship, the issue is now fixed. Can you test again? i also added a loading indicator to the signUp page.

**Comment by techno-disaster** [comment]
@anitab-org/qa-team can anyone give this a quick test again? Thank You

**Review by robotjellyzone — DISMISSED** [review]
Hi @Techno-Disaster i have tested the required feature & the relation page , everything is now working fine 

great  👍 

testing_pr_2

File: lib/widgets/loading_indicator.dart [hunk 6]
```diff
@@ -9,3 +9,11 @@ class LoadingIndicator extends StatelessWidget {
     );
   }
 }
+
+Future<void> showProgressIndicator(BuildContext context) {
```
Comment (bartekpacia) [inline lib/widgets/loading_indicator.dart:13]: `topcontext` -> `topContext` ;)

**Review by bartekpacia — DISMISSED** [review]
Nice job :)

File: lib/widgets/loading_indicator.dart (hunk 6 above)
Comment (bartekpacia) [reply lib/widgets/loading_indicator.dart:13]: I marked wrong lines, but you surely understand what I mean

Comment (techno-disaster) [reply lib/widgets/loading_indicator.dart:13]: got it, will get it done asap :)

Comment (robotjellyzone) [reply lib/widgets/loading_indicator.dart:13]: > `topcontext` -> `topContext` ;)

camel casing good catch 👍

Comment (techno-disaster) [reply lib/widgets/loading_indicator.dart:13]: Done

**Comment by techno-disaster** [comment]
@anitab-org/mentorship-flutter-maintainers Can you take a quick look at this? The PR has been tested (all OK) just needs a quick review now

**Review by isabelcosta — APPROVED** [review]
Knowing very little about flutter, I am approving because this functionality seems to work.
@Techno-Disaster great work.
I am counting on other reviewers for code quality check. @bartekpacia you have to re-review after recent code changes.

**Comment by techno-disaster** [comment]
@bartekpacia  could you check this again? previous review git dismissed due to the ```topcontext``` > ```topContext``` change

**Review by bartekpacia — APPROVED** [review]
thanks for using the `camelCase`, good job!

**Pull request merged**